APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -575,113 +735,148 @@ - (NSArray *)retrieveResultSummaryFromResults:(NSSet *)results latestOnly:(BOOL)
     return allResultSummaries;
 }
 
-- (void)groupDatasetByDay
+- (void)groupByPeriod:(APHTimelineGroups)period
 {
-    NSMutableArray *groupedDataset = [NSMutableArray array];
-    NSArray *days = [self.dataPoints valueForKeyPath:@"@distinctUnionOfObjects.datasetDateKey"];
+    NSString *periodKeyPath = nil;
+    
+    switch (period) {
+        case APHTimelineGroupWeek:
+            periodKeyPath = kDatasetGroupByWeek;
+            break;
+        case APHTimelineGroupMonth:
+            periodKeyPath = kDatasetGroupByMonth;
+            break;
+        case APHTimelineGroupYear:
+            periodKeyPath = kDatasetGroupByYear;
+            break;
+        default:
+            periodKeyPath = kDatasetGroupByDay;
+            break;
+    }
+    
+    // Group the dataset based on the provided period.
+    NSDictionary *groupedDataset = [self groupByKeyPath:periodKeyPath dataset:self.dataPoints];
     
-    for (NSString *day in days) {
-        NSMutableDictionary *entry = [NSMutableDictionary dictionary];
-        [entry setObject:day forKey:kDatasetDateKey];
+    // Summarize data for each group.
+    NSDictionary *summarizedDataset = [self summarizeDataset:groupedDataset period:period];
+    NSArray *sortedSummaryKeys = [summarizedDataset.allKeys sortedArrayUsingSelector:@selector(compare:)];
+    
+    NSMutableArray *sortedSummarizedDataset = [NSMutableArray new];
+    
+    for (id key in sortedSummaryKeys) {
+        [sortedSummarizedDataset addObjectsFromArray:summarizedDataset[key]];
+    }
+    
+    [self.dataPoints removeAllObjects];
+    [self.dataPoints addObjectsFromArray:sortedSummarizedDataset];
+}
+
+- (NSDictionary *)summarizeDataset:(NSDictionary *)dataset period:(APHTimelineGroups)period
+{
+    NSMutableDictionary *summarizedDataset = [NSMutableDictionary new];
+    NSArray *keys = [dataset allKeys];
+    
+    for (id key in keys) {
+        NSArray *elements = dataset[key];
+        NSDictionary *rawData = nil;
         
-        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"(%K = %@) and (%K <> %@)", kDatasetDateKey, day, kDatasetValueKey, @(NSNotFound)];
-        NSArray *groupItems = [self.dataPoints filteredArrayUsingPredicate:predicate];
+        if (period == APHTimelineGroupForInsights) {
+            // We only need the latest for the raw data.
+            // Since our filtered array is sorted in a desending order
+            // we can take the first object.
+            NSSortDescriptor *sortByDate = [[NSSortDescriptor alloc] initWithKey:kDatasetDateKey ascending:NO];
+            NSArray *latestRawData = [elements sortedArrayUsingDescriptors:@[sortByDate]];
```

Comment:
It doesn't the comment had a typo, the incoming elements array is sorted in ascending order. The code has been revised and now we are taking the last object in the elements array, since that would be the latest data point.